### PR TITLE
Check power commands

### PR DESCRIPTION
**Pull request by Gregg140**
This stand along program will check the ACIS commands in the backstop file for proper timing.  The timing rules can be read in Notes.txt.

This program is called by the ACIS Ops LR software as the very last thing it does before exiting.

### CODE REVIEW

**Comment by Gregg140** [comment]
Here are some screen shots of the TKDIFF's done MAR0518 load (modified to induce errors_
error_mar0518_power_command_violations_1_and_2
error_mar0518_power_command_violations_3
error_mar0518_power_command_violations_4
error_mar0518_power_command_violations_5
error_mar0518_power_command_violations_6




:

**Review by richardjedgar — APPROVED** [review]
Looks good to me.

File: CHECK_POWER_COMMANDS/System_State_Class.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+ 
+    #---------------------------------------------------------------------------
+    #
+    # Method: Print System State: Print out the present values of the 
+    #                             system state parameters
+    #
+    #---------------------------------------------------------------------------
+    def print_system_state(self,):
```
Comment (jzuhone) [inline CHECK_POWER_COMMANDS/System_State_Class.py]: This won't run in Python 3 when the ska Python switches to that, so maybe we should do the `from __future__ import print_function` at the top and then make these calls to `print()`.

File: CHECK_POWER_COMMANDS/Rulesets.py [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+    if cmd_entry['packet_or_cmd'] == 'OORMPEN':
+        # You are done with the perigee passage - 
+        # Record that system_state
+        present_state['science_run_exec'] = False
+        present_state['perigee_passage'] = False
+        rules_fired.append('ORB/CMD_SW Rule 3 - Exiting Perigee Passage')
+ 
+
+    # Return the system_state and the rules fired list
+    return ( rules_fired)
```
Comment (jzuhone) [inline CHECK_POWER_COMMANDS/Rulesets.py]: these parentheses are extraneous

File: CHECK_POWER_COMMANDS/Rulesets.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+        violation['vio_time'] =  ACISPKT_state['time_cmd']
+        violation['vio_rule'] = 'ACISPKT Rule #6 - ERROR ALL FEPS off and it is  1 hour past WSPOW00000 time'
+        violations_list.append(violation )
+        # Record which rule fired
+        rules_fired.append('ACISPKT Rule #6 - ERROR 3 FEPS off and it is  1 hour past WSPOW00000 time')
+
+    
+
+    # Return the state and the rules fired list
+    return ( rules_fired, violations_list)
```
Comment (jzuhone) [inline CHECK_POWER_COMMANDS/Rulesets.py]: extraneous parentheses

File: CHECK_POWER_COMMANDS/Check_Power_Cmds.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        # Save the present system state in another variable
+        last_state = dict(system_state.state)
+        # Run the ACISPKT ruleset once
+        new_rules_fired, violations_list = rules.ACISPKT_rules(eachpacket,
+                                                               last_state,
+                                                               system_state,
+                                                               bfc,
+                                                               violations_list)
+        # Append all the rules that may have fired
+        if new_rules_fired:
```
Comment (jzuhone) [inline CHECK_POWER_COMMANDS/Check_Power_Cmds.py]: I think that it would be a bit clearer what's happening here if we checked the length of the `new_rules_fired` list, so:

```python
if len(new_rules_fired) > 0:
    ...
```

File: CHECK_POWER_COMMANDS/Check_Power_Cmds.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+            # Set the last state to the present state
+            last_state = dict(system_state.state)
+            # Run the ACISPKT ruleset again
+            new_rules_fired, violations_list = rules.ACISPKT_rules(eachpacket,
+                                                                   last_state,
+                                                                   system_state,
+                                                                   bfc,
+                                                                   violations_list)
+            # Append all the rules that may have fired
+            if new_rules_fired:
```
Comment (jzuhone) [inline CHECK_POWER_COMMANDS/Check_Power_Cmds.py]: Same comment as above about checking the length

File: CHECK_POWER_COMMANDS/Check_Power_Cmds.py [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+                all_rules_fired.append(list(new_rules_fired))
+
+        # Store the command you are assessing as the previous command
+        bfc.write_previous_ACISPKT_cmd(eachpacket)
+                
+    else:
+        # Else it's not an acis packet so run the CMD/ORB rule set
+        rules_fired = rules.ORB_CMD_SW_rule_set(eachpacket, system_state.state, violations_list)
+        # If any rules fired, append them to the list
+        if rules_fired:
```
Comment (jzuhone) [inline CHECK_POWER_COMMANDS/Check_Power_Cmds.py]: Same comment as above on checking the length

**Review by jzuhone — CHANGES_REQUESTED** [review]
Ok, I had a look at this and essentially all of the comments I made are style changes. There are a few which I think would make the code clearer. But otherwise, looks good.

File: CHECK_POWER_COMMANDS/Backstop_File_Class.py [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+#                        generated by the Check_Power_cmds program. 
+#
+################################################################################
+import pprint
+import re
+import numpy as np
+
+from Chandra.Time import DateTime
+
+class Backstop_File_Object:
```
[the remark this answers is not in the document]
Comment (Gregg140) [reply CHECK_POWER_COMMANDS/Backstop_File_Class.py]: That makes it SO MUCH harder to read - and easy to confuse with another slightly different name if there was one -  but I'll make this change.

File: CHECK_POWER_COMMANDS/Backstop_File_Class.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+                # the date, time, insert the word ACISPKT, and the mnemonic
+                self.system_packets = np.r_[self.system_packets,
+                                     np.array( [ ( split_line[0],
+                                                   DateTime(packet_time).secs,
+                                                   'ACISPKT',
+                                                   cmd) ],
+                                               dtype = self.ACISPKT_dtype) ]
+        
+            # Next check if the line is one of the perigee Passage indicators
+            if [True for pp_ind in pp_indicators if (pp_ind in eachline)]:
```
[the remark this answers is not in the document]
Comment (Gregg140) [reply CHECK_POWER_COMMANDS/Backstop_File_Class.py]: Yes I like the accepted answer better. Implemented

File: CHECK_POWER_COMMANDS/Backstop_File_Class.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+        # Open the output file
+        outfile = open(ACISPKT_file, 'w')
+        
+        # Open the load review text file
+        infile = open(backstop_file, 'r')
+        # Read in each line. If the line contains 'ACISPKT" then
+        # write it to the output file
+        for eachline in infile:
+            # If it's an ACISPKT line grab it without question
+            if ('ACISPKT' in eachline):
```
[the remark this answers is not in the document]
Comment (Gregg140) [reply CHECK_POWER_COMMANDS/Backstop_File_Class.py]: Extraneous but it doesn't cause a problem nor any time delay and I much prefer having the IF antecedents surrounded by parens

File: CHECK_POWER_COMMANDS/Rulesets.py (hunk 2 above)
Comment (Gregg140) [reply CHECK_POWER_COMMANDS/Rulesets.py]: See above

File: CHECK_POWER_COMMANDS/Rulesets.py (hunk 3 above)
Comment (Gregg140) [reply CHECK_POWER_COMMANDS/Rulesets.py]: See Above

File: CHECK_POWER_COMMANDS/Rulesets.py (hunk 2 above)
Comment (Gregg140) [reply CHECK_POWER_COMMANDS/Rulesets.py]: on the returns I'll remove the parens

File: CHECK_POWER_COMMANDS/Check_Power_Cmds.py (hunk 4 above)
Comment (Gregg140) [reply CHECK_POWER_COMMANDS/Check_Power_Cmds.py]: implemented

File: CHECK_POWER_COMMANDS/Check_Power_Cmds.py (hunk 5 above)
Comment (Gregg140) [reply CHECK_POWER_COMMANDS/Check_Power_Cmds.py]: implemented

File: CHECK_POWER_COMMANDS/Check_Power_Cmds.py (hunk 4 above)
Comment (Gregg140) [reply CHECK_POWER_COMMANDS/Check_Power_Cmds.py]: actually NOT implemented.

File: CHECK_POWER_COMMANDS/Check_Power_Cmds.py (hunk 5 above)
Comment (Gregg140) [reply CHECK_POWER_COMMANDS/Check_Power_Cmds.py]: not implemented

**Comment by Gregg140** [comment]
Ok I made several of the changes John suggested. Others I did not. Please review.

Thanks!

**Comment by jzuhone** [comment]
@Gregg140 one last quick question--I assume you re-ran after these changes and got the same answers as before?

**Comment by Gregg140** [comment]
Sorry yeah I meant to say that too,. I re-ran all regression test cases and the results were identical.

**Review by jzuhone — APPROVED** [review]

File: CHECK_POWER_COMMANDS/Check_Power_Cmds.py (hunk 4 above)
Comment (taldcroft) [reply CHECK_POWER_COMMANDS/Check_Power_Cmds.py]: Just my 2c that I agree with @jzuhone on this general convention as it makes for more robust software.  For instance if `new_rules_fired` were a function parameter and someone passed in `None` then the test would happily produce `False` and move on.  Likewise if someone passed in a numpy array this test would do a variety of things (exception, warning, or bool) depending on the details of the array, details of numpy / Python version, etc etc.  

Instead if you use `len()` then you really know what you are testing.

tl;dr - explicit is better than implicit

Comment (Gregg140) [reply CHECK_POWER_COMMANDS/Check_Power_Cmds.py]: While I tend to agree, you do realize, of course that this violates PEP-8:

For sequences, (strings, lists, tuples), use the fact that empty sequences are false.

Yes: if not seq:
     if seq:

No: if len(seq):
    if not len(seq):
